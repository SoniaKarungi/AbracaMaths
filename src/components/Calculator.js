import React, { useState, useEffect } from 'react';
import Button from './Button';
import calculate from '../logic/calculate';

import '../css/calculator.css';

function Calculator() {
  const [calcObject, setCalcObject] = useState({
    total: null,
    next: null,
    operation: null,
  });

  useEffect(() => {
    const responseElement = document.getElementById('results');
    const { total, next } = calcObject;
    responseElement.textContent = next || total;
  });

  const handleClick = (buttonName) => {
    setCalcObject((prevState) => {
      const newCalculatorObject = {
        ...prevState,
        ...calculate(prevState, buttonName),
      };
      return newCalculatorObject;
    });
  };

  return (
    <div className="container">
      <h2>Let&#39;s do some maths!</h2>
      <div className="calc-container">
        <div className="screen" id="results" data-testid="results">
          Results Here...
        </div>
        <Button
          cssClassName="btn"
          buttonText="AC"
          onButtonClick={handleClick}
        />
        <Button
          cssClassName="btn"
          buttonText="+/-"
          onButtonClick={handleClick}
        />
        <Button cssClassName="btn" buttonText="%" onButtonClick={handleClick} />
        <Button
          cssClassName="btn"
          buttonText="รท"
          orangeClassName="orange"
          onButtonClick={handleClick}
        />
        <Button cssClassName="btn" buttonText="7" onButtonClick={handleClick} />
        <Button cssClassName="btn" buttonText="8" onButtonClick={handleClick} />
        <Button cssClassName="btn" buttonText="9" onButtonClick={handleClick} />
        <Button
          cssClassName="btn"
          buttonText="x"
          orangeClassName="orange"
          onButtonClick={handleClick}
        />
        <Button cssClassName="btn" buttonText="4" onButtonClick={handleClick} />
        <Button cssClassName="btn" buttonText="5" onButtonClick={handleClick} />
        <Button cssClassName="btn" buttonText="6" onButtonClick={handleClick} />
        <Button
          cssClassName="btn"
          buttonText="-"
          orangeClassName="orange"
          onButtonClick={handleClick}
        />
        <Button cssClassName="btn" buttonText="1" onButtonClick={handleClick} />
        <Button cssClassName="btn" buttonText="2" onButtonClick={handleClick} />
        <Button cssClassName="btn" buttonText="3" onButtonClick={handleClick} />
        <Button
          cssClassName="btn"
          buttonText="+"
          orangeClassName="orange"
          onButtonClick={handleClick}
        />
        <Button
          cssClassName="btn"
          buttonText="0"
          orangeClassName="zero"
          onButtonClick={handleClick}
        />
        <Button cssClassName="btn" buttonText="." onButtonClick={handleClick} />
        <Button
          cssClassName="btn"
          buttonText="="
          orangeClassName="orange"
          onButtonClick={handleClick}
        />
      </div>
    </div>
  );
}

export default Calculator;
