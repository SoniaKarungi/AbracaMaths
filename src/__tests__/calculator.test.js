import React from 'react';
import { fireEvent, render, screen } from '@testing-library/react';
import renderer from 'react-test-renderer';
import Calculator from '../components/Calculator';

describe('Calculator', () => {
  it('renders correctly', () => {
    const tree = renderer.create(<Calculator />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('contains exactly 19 buttons', () => {
    render(<Calculator />);

    const buttons = screen.getAllByRole('button');

    expect(buttons.length).toBe(19);
  });

  it('returns the correct result after the 2 button is clicked', () => {
    render(<Calculator />);
    const resultElement = screen.getByTestId(/results/);
    const twoButton = screen.getByText(/2/);

    fireEvent.click(twoButton);

    expect(resultElement).toHaveTextContent(2);
  });

  it('returns the correct result for addition operation', () => {
    render(<Calculator />);
    const resultElement = screen.getByTestId(/results/);
    const twoButton = screen.getByText(/2/);
    const plusButton = screen.getByText(/^[+]$/);
    const equalButton = screen.getByText(/=/);

    fireEvent.click(twoButton);
    fireEvent.click(plusButton);
    fireEvent.click(twoButton);
    fireEvent.click(equalButton);

    expect(resultElement).toHaveTextContent(4);
  });

  it('returns the correct result for subtraction operation', () => {
    render(<Calculator />);
    const resultElement = screen.getByTestId(/results/);
    const twoButton = screen.getByText(/2/);
    const plusButton = screen.getByText(/^[-]$/);
    const equalButton = screen.getByText(/=/);

    fireEvent.click(twoButton);
    fireEvent.click(plusButton);
    fireEvent.click(twoButton);
    fireEvent.click(equalButton);

    expect(resultElement).toHaveTextContent(0);
  });

  it('returns an error after division by 0 operation', () => {
    render(<Calculator />);
    const resultElement = screen.getByTestId(/results/);
    const threeButton = screen.getByText(/3/);
    const zeroButton = screen.getByText(/0/);
    const dividedByButton = screen.getByText(/^[รท]$/);
    const equalButton = screen.getByText(/=/);

    fireEvent.click(threeButton);
    fireEvent.click(dividedByButton);
    fireEvent.click(zeroButton);
    fireEvent.click(equalButton);

    expect(resultElement).toHaveTextContent("Can't divide by 0.");
  });

  it('returns an error after a modulo by 0 operation', () => {
    render(<Calculator />);
    const resultElement = screen.getByTestId(/results/);
    const threeButton = screen.getByText(/3/);
    const zeroButton = screen.getByText(/0/);
    const modeButton = screen.getByText(/^[%]$/);
    const equalButton = screen.getByText(/=/);

    fireEvent.click(threeButton);
    fireEvent.click(modeButton);
    fireEvent.click(zeroButton);
    fireEvent.click(equalButton);

    expect(resultElement).toHaveTextContent(
      "Can't find modulo as can't divide by 0.",
    );
  });
});
