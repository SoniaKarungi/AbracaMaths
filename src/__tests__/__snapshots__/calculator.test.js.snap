// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Calculator renders correctly 1`] = `
<div
  className="container"
>
  <h2>
    Let's do some maths!
  </h2>
  <div
    className="calc-container"
  >
    <div
      className="screen"
      data-testid="results"
      id="results"
    >
      Results Here...
    </div>
    <button
      className=" btn"
      onClick={[Function]}
      type="button"
    >
      AC
    </button>
    <button
      className=" btn"
      onClick={[Function]}
      type="button"
    >
      +/-
    </button>
    <button
      className=" btn"
      onClick={[Function]}
      type="button"
    >
      %
    </button>
    <button
      className="orange btn"
      onClick={[Function]}
      type="button"
    >
      รท
    </button>
    <button
      className=" btn"
      onClick={[Function]}
      type="button"
    >
      7
    </button>
    <button
      className=" btn"
      onClick={[Function]}
      type="button"
    >
      8
    </button>
    <button
      className=" btn"
      onClick={[Function]}
      type="button"
    >
      9
    </button>
    <button
      className="orange btn"
      onClick={[Function]}
      type="button"
    >
      x
    </button>
    <button
      className=" btn"
      onClick={[Function]}
      type="button"
    >
      4
    </button>
    <button
      className=" btn"
      onClick={[Function]}
      type="button"
    >
      5
    </button>
    <button
      className=" btn"
      onClick={[Function]}
      type="button"
    >
      6
    </button>
    <button
      className="orange btn"
      onClick={[Function]}
      type="button"
    >
      -
    </button>
    <button
      className=" btn"
      onClick={[Function]}
      type="button"
    >
      1
    </button>
    <button
      className=" btn"
      onClick={[Function]}
      type="button"
    >
      2
    </button>
    <button
      className=" btn"
      onClick={[Function]}
      type="button"
    >
      3
    </button>
    <button
      className="orange btn"
      onClick={[Function]}
      type="button"
    >
      +
    </button>
    <button
      className="zero btn"
      onClick={[Function]}
      type="button"
    >
      0
    </button>
    <button
      className=" btn"
      onClick={[Function]}
      type="button"
    >
      .
    </button>
    <button
      className="orange btn"
      onClick={[Function]}
      type="button"
    >
      =
    </button>
  </div>
</div>
`;
